basicsr/archs/hifacegan_arch.py

```diff
@@ -0,0 +1,265 @@
+            x = self.ups[i](x, guide_list[4 + i])
+
+        x = self.to_rgbs[phase - 1](F.leaky_relu(x, 2e-1))
+        x = torch.tanh(x)
+
+        return x
+
+
+@ARCH_REGISTRY.register()
+class HiFaceGAN(SPADEGenerator):
+    """ HiFaceGAN: SPADEGenerator with a learnable feature encoder
+        Current encoder design: LIPEncoder"""
+
+    def __init__(self,
+                 num_in_ch=3,
+                 num_feat=64,
+                 use_vae=False,
+                 z_dim=256,
+                 crop_size=512,
+                 norm_g='spectralspadesyncbatch3x3',
+                 is_train=True,
+                 init_train_phase=3):
+
+        super().__init__(num_in_ch, num_feat, use_vae, z_dim, crop_size, norm_g, is_train, init_train_phase)
+        self.lip_encoder = LIPEncoder(num_in_ch, num_feat, self.sw, self.sh, self.scale_ratio)
+
+    def encode(self, input_tensor):
+        return self.lip_encoder(input_tensor)
+
+
+@ARCH_REGISTRY.register()
+class HiFaceGANDiscriminator(BaseNetwork):
+    """ Inspired by pix2pixHD multiscale discriminator.
+
+    Args:
+        num_in_ch (int): Channel number of inputs. Default: 3.
+        num_out_ch (int): Channel number of outputs. Default: 3.
+        conditional_D (bool): Whether use conditional discriminator.
+            Default: True.
+        num_D (int): Number of Multiscale discriminators. Default: 3.
+        n_layers_D (int): Number of downsample layers in each D. Default: 4.
+        num_feat (int): Channel number of base intermediate features.
+            Default: 64.
+        norm_D (str): String to determine normalization layers in D.
+            Choices: [spectral][instance/batch/syncbatch]
+            Default: 'spectralinstance'.
+        keep_features (bool): Keep intermediate features for matching loss, etc.
+            Default: True.
+    """
+
+    def __init__(self,
+                 num_in_ch=3,
+                 num_out_ch=3,
+                 conditional_D=True,
+                 num_D=2,
+                 n_layers_D=4,
+                 num_feat=64,
+                 norm_D='spectralinstance',
+                 keep_features=True):
+        super().__init__()
+        self.num_D = num_D
+
+        input_nc = num_in_ch
+        if conditional_D:
+            input_nc += num_out_ch
+
+        for i in range(num_D):
+            subnet_D = NLayerDiscriminator(input_nc, n_layers_D, num_feat, norm_D, keep_features)
+            self.add_module(f'discriminator_{i}', subnet_D)
+
+    def downsample(self, x):
+        return F.avg_pool2d(x, kernel_size=3, stride=2, padding=[1, 1], count_include_pad=False)
+
+    # Returns list of lists of discriminator outputs.
+    # The final result is of size opt.num_D x opt.n_layers_D
+    def forward(self, x):
+        result = []
+        for _, D in self.named_children():
+            out = D(x)
+            result.append(out)
+            x = self.downsample(x)
+
+        return result
+
+
+@ARCH_REGISTRY.register()
```

Comment:
only add this line `@ARCH_REGISTRY.register()` when we will use it in the yml comfig file. 
It seems that NLayerDiscriminatoris used in the `HiFaceGANDiscriminator` class